queue/queue.mbt

```diff
@@ -0,0 +1,364 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+struct Cell[T] {
+  content : T
+  mut next : Option[Cell[T]]
+}
+
+pub fn op_equal[T : Eq](self : Cell[T], other : Cell[T]) -> Bool {
+  self.content == other.content
+}
+
+struct Queue[T] {
+  mut length : Int
+  mut tail : Option[Cell[T]]
+}
+
+/// Creates a new empty queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::create()
+/// ```
+pub fn Queue::new[T]() -> Queue[T] {
+  { length: 0, tail: None }
+}
+
+test "new" {
+  let queue : Queue[Int] = new()
+  @assertion.assert_eq(queue.length, 0)?
+  @assertion.assert_true(queue.tail == None)?
+}
+
+/// Creates a new queue from an array.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// ```
+pub fn Queue::from_array[T](array : Array[T]) -> Queue[T] {
+  let queue : Queue[T] = new()
```

Comment:
if we build queue from scratch, we can make it significantly faster without using add?